File: SmallLabyWpfPlayer/Model/ClientModel.cs
using SmallLabyWpfPlayer.ServiceRefSmallLaby;
using System.Collections.Generic;
using System.Linq;

namespace SmallLabyWpfPlayer
{
    public class ClientModel
    {
        public enum Terrain
        {
            Grass,
            Mountain,
            Water,
            Forest,
            Unknown,
        }
        public enum ItemType
        {
            Gold,
            Bonus,
            Exit,
            Unknown,
        }

        public enum MovementStrategy
        {
            MoveUp,
            MoveDown,
            MoveLeft,
            MoveRight,
            DoNotMove,
            MoveRandom,
        }
        public struct Player
        {
            public int X;
            public int Y;
            public string Name;
            public bool Me;
        }
        public struct Monster
        {
            public int X;
            public int Y;
        }
        public struct Field
        {
            public int X;
            public int Y;
            public Terrain Terrain;
        }
        public struct Item
        {
            public int X;
            public int Y;
            public ItemType Type;
        }

        private SmallLabyClient m_client;
        private int m_player_id;
        private string m_player_name;
        private TerrainType[] m_map;

      public int MapWidth { get; }
      public int MapHeight { get; }

        public ClientModel()
        {
            m_client = new SmallLabyClient();
            m_map = m_client.GetMap();
            MapWidth = m_client.GetMapWidth();
            MapHeight = m_client.GetMapHeight();
            IsConnected = false;
        }

        public void Connect(string player_name)
        {
            m_player_id = m_client.AddPlayer(player_name);
            m_player_name = player_name;
            IsConnected = true;
        }

        public void Disconnect()
        {
            if (IsConnected)
            {
                m_client.RemovePlayer(m_player_id);
                IsConnected = false;
            }
        }

        public IEnumerable<Player> GetPlayers()
        {
            return m_client.GetPlayers().Select(p =>
            new Player
            {
                X = p.X,
                Y = p.Y,
                Name = (p.Id == m_player_id) ? m_player_name : "enemy",
                Me = p.Id == m_player_id
            });
        }

        public IEnumerable<Monster> GetMonsters()
        {
            return m_client.GetMonsters().Select(m =>
            new Monster
            {
                X = m.X,
                Y = m.Y,
            });
        }

        public IEnumerable<Item> GetItems()
        {
            foreach (var item_info in m_client.GetItems())
            {
                ItemType type = ItemType.Unknown;
                switch (item_info.Item)
                {
                    case ServiceRefSmallLaby.Item.Gold:
                        type = ItemType.Gold;
                        break;
                    case ServiceRefSmallLaby.Item.Bonus:
                        type = ItemType.Bonus;
                        break;
                    case ServiceRefSmallLaby.Item.Exit:
                        type = ItemType.Exit;
                        break;
                }
                yield return new Item
                {
                    X = item_info.X,
                    Y = item_info.Y,
                    Type = type,
                };
            }
        }
        public IEnumerable<Field> GetMap()
        {
            for (int y = 0; y < MapHeight; y++)
            {
                for (int x = 0; x < MapWidth; x++)
                {
                    var field = m_map[y * MapWidth + x];

                    Terrain terrain = Terrain.Unknown;
                    switch (field)
                    {
                        case TerrainType.Grass:
                            terrain = Terrain.Grass;
                            break;
                        case TerrainType.Mountain:
                            terrain = Terrain.Mountain;
                            break;
                        case TerrainType.Water:
                            terrain = Terrain.Water;
                            break;
                        case TerrainType.Forest:
                            terrain = Terrain.Forest;
                            break;
                    }
                    yield return new Field { X = x, Y = y, Terrain = terrain };
                }
            }
        }

        public void SetMovementStrategy(MovementStrategy strategy)
        {
            if (!IsConnected)
                return;
            switch (strategy)
            {
                case MovementStrategy.MoveLeft:
                    m_client.SetMovementStrategy(m_player_id,
                        ServiceRefSmallLaby.MovementStrategy.MoveLeft);
                    break;
                case MovementStrategy.MoveRight:
                    m_client.SetMovementStrategy(m_player_id,
                        ServiceRefSmallLaby.MovementStrategy.MoveRight);
                    break;
                case MovementStrategy.MoveUp:
                    m_client.SetMovementStrategy(m_player_id,
                        ServiceRefSmallLaby.MovementStrategy.MoveUp);
                    break;
                case MovementStrategy.MoveDown:
                    m_client.SetMovementStrategy(m_player_id,
                        ServiceRefSmallLaby.MovementStrategy.MoveDown);
                    break;
                case MovementStrategy.DoNotMove:
                    m_client.SetMovementStrategy(m_player_id,
                        ServiceRefSmallLaby.MovementStrategy.StandStill);
                    break;
                case MovementStrategy.MoveRandom:
                    m_client.SetMovementStrategy(m_player_id,
                        ServiceRefSmallLaby.MovementStrategy.RandomDirection);
                    break;
            }
        }

        public bool IsConnected { get; private set; }

        public int PlayerGold => IsConnected ? m_client.GetGold(m_player_id) : 0;

        public double PlayerSpeed => IsConnected ? m_client.GetSpeed(m_player_id) : 0;

    }
}

File: SmallLabySfml/ClientModel.cpp
#include "stdafx.h"
#include "ClientModel.h"

#include <vcclr.h>
#include <string>
using namespace System;
using namespace SmallLabyUser;
//using namespace SmallLabyUser::ServiceRefSmallLaby;

struct ClientModel::PrivateModel
{
  gcroot<ServiceRefSmallLaby::SmallLabyClient^> client;
  PrivateModel()
  {
    client = gcnew ServiceRefSmallLaby::SmallLabyClient();
  }
  ~PrivateModel()
  {
    client->Close();
  }
};

ClientModel::ClientModel()
{
  _model_impl = std::make_unique<PrivateModel>();
}


ClientModel::~ClientModel()
{
}

int ClientModel::GetWidth()
{
  return _model_impl->client->GetMapWidth();
}

int ClientModel::GetHeight()
{
  return _model_impl->client->GetMapHeight();
}

std::vector<std::vector<Terrain>> ClientModel::GetMap()
{
  array<ServiceRefSmallLaby::TerrainType>^ map = _model_impl->client->GetMap();
  int map_height = GetHeight();
  int map_width = GetWidth();
  std::vector<std::vector<Terrain>> terrain_map;
  terrain_map.resize(map_height);
  for (int y = 0; y < map_height; y++)
  {
    terrain_map[y].resize(map_width);
    for (int x = 0; x < map_width; x++)
    {
      ServiceRefSmallLaby::TerrainType field = map[y * map_width + x];
      switch (field)
      {
      case ServiceRefSmallLaby::TerrainType::Mountain:
        terrain_map[y][x] = Terrain::Wall;
        break;
      default:
        terrain_map[y][x] = Terrain::Road;
        break;
      }
    }
  }
  return terrain_map;
}

int ClientModel::AddPlayer(std::string name)
{
  return _model_impl->client->AddPlayer("alex");
}

void ClientModel::SetMoveStrategy(int player_id, MoveStrategy movestrategy)
{
  ServiceRefSmallLaby::MovementStrategy movementstrategy;
  switch (movestrategy)
  {
  case MoveStrategy::MoveLeft:
    movementstrategy = ServiceRefSmallLaby::MovementStrategy::MoveLeft;
    break;
  case MoveStrategy::MoveRight:
    movementstrategy = ServiceRefSmallLaby::MovementStrategy::MoveRight;
    break;
  case MoveStrategy::MoveUp:
    movementstrategy = ServiceRefSmallLaby::MovementStrategy::MoveUp;
    break;
  case MoveStrategy::MoveDown:
    movementstrategy = ServiceRefSmallLaby::MovementStrategy::MoveDown;
    break;
  default:
    break;
  }
  _model_impl->client->SetMovementStrategy(player_id, movementstrategy);
}

void ClientModel::RemovePlayer(int player_id)
{
  _model_impl->client->RemovePlayer(player_id);
}

void ClientModel::Close()
{
  _model_impl->client->Close();
}

std::vector<Player> ClientModel::GetPlayers()
{
  array<ServiceRefSmallLaby::PlayerInfo>^ players = _model_impl->client->GetPlayers();
  std::vector<Player> playersv;
  for (int i = 0; i < players->GetLength(0); i++)
  {
    Player player;
    player.X = players[i].X;
    player.Y = players[i].Y;
    player.Id = players[i].Id;
    playersv.push_back(player);
  }
  return playersv;
}

std::vector<Monster> ClientModel::GetMonsters()
{
  array<ServiceRefSmallLaby::MonsterInfo>^ monsters = _model_impl->client->GetMonsters();
  std::vector<Monster> monstersv;
  for (int i = 0; i < monsters->GetLength(0); i++)
  {
    Monster monster;
    monster.X = monsters[i].X;
    monster.Y = monsters[i].Y;
    monstersv.push_back(monster);
  }
  return monstersv;
}